### ISSUE TITLE
Feature Request (or Guidance Request): Frame-by-Frame Processing with ONNX Export in Flash Linear Attention Model

### ISSUE THREAD

**Issue by LarocheC** [sustcsonglin/flash-linear-attention#79]
### Feature request

I am working with flash linear attention (FLA) model and aiming to process data in a frame-by-frame manner with [batch, frame, feature] input format. Additionally, I am looking to export the model to ONNX for broader deployment options. However, I'm not entirely familiar with all aspects of the repository, so it’s possible that I might have missed some existing functionality or best practices for achieving this.

### Motivation

To use the FLA models in other applications, I would need to process data on a frame-by-frame basis while maintaining states across frames (for real-time applications). I’ve also found ONNX compatibility interesting for deployment but am unsure if the current implementation of the FLA model fully supports this export with state preservation between frames.

### Your contribution

Here is a quick script that shows what I'm looking into. 
```import fla
import torch 

batch_size, num_heads, seq_len, hidden_size = 6, 4, 128, 256
device, dtype = 'cuda', torch.float32
RWKV6 = fla.layers.RWKV6Attention(hidden_size=hidden_size, num_heads=num_heads).to(device=device, dtype=dtype)

x = torch.randn(batch_size, seq_len, hidden_size).to(device=device, dtype=dtype)
y, *_ = RWKV6(x)

from fla.models.utils import Cache
data = torch.randn(batch_size, seq_len, hidden_size).to(device=device, dtype=dtype)
initial_cache = Cache(seen_tokens=0)

past_key_values = initial_cache  # Use initialized cache as starting states
outputs_frame_by_frame = []

for frame_idx in range(seq_len):
    frame_data = data[:, frame_idx:frame_idx+1, :]  # Shape: [batch, 1, feature]
    frame_output, _, past_key_values = RWKV6(
        frame_data,
        past_key_values=past_key_values,
        use_cache=True
    )
    past_key_values = past_key_values  # Update the cache for the next frame
    outputs_frame_by_frame.append(frame_output)

final_output_frame_by_frame = torch.cat(outputs_frame_by_frame, dim=1)  # Shape: [batch, total_frames, hidden_size]

final_output_all_at_once, _, _ = RWKV6(
    data,
    past_key_values=None,  # Start with the same initialized states
    use_cache=False
)

difference = torch.abs(final_output_frame_by_frame - final_output_all_at_once)
max_difference = torch.max(difference)
mean_difference = torch.mean(difference)

print(f"Max difference: {max_difference.item()}")
print(f"Mean difference: {mean_difference.item()}")

**Comment by yzhangcs** [comment]
@LarocheC Hello, if u do need to use fla in a frame-by-frame manner, the above code is ok for me. I think the above logic is essentially what the hf api `model.generate` involves.
Pay careful attention to `past_key_values ` management and you may need to preallocate the storage for all the outputs for faster decoding.

Regarding ONNX, I'm not familiar with this before 😢 . Are u interested in making any PRs for it? `fla` models are entirely compatible with hf-style.

**Comment by LarocheC** [comment]
Hi @yzhangcs 

Thank you for the response and the guidance!

I’m unfamiliar with the `model.generate` API, but before I proceed, I’ll try to use it to reproduce the frame-by-frame processing you suggested.

Regarding ONNX, I do have some experience converting RNN models to ONNX and handling states externally (which is necessary for real-time applications), so I’ll look into whether a similar approach could be applied here for the FLA model. 

Do you think this would be a relevant feature for the repository? I’d be open to exploring it further if it could add value.

Thanks again for your help!

**Comment by yzhangcs** [comment]
@LarocheC your contributions are welcome if you are interested in!
You can add scripts [here](https://github.com/sustcsonglin/flash-linear-attention/tree/main/utils).
